backends/magma/ceed-magma-basis.c

```diff
@@ -499,39 +608,127 @@ int CeedBasisCreateTensorH1_Magma(CeedInt dim, CeedInt P1d, CeedInt Q1d,
-  if (ncomp > 3)
-    // LCOV_EXCL_START
-    return CeedError(ceed, CEED_ERROR_BACKEND,
-                     "Magma backend does not support tensor bases with more than 3 components");
-  // LCOV_EXCL_STOP
-  if (P1d > 10)
-    // LCOV_EXCL_START
-    return CeedError(ceed, CEED_ERROR_BACKEND,
-                     "Magma backend does not support tensor bases with more than 10 nodes in each dimension");
-  // LCOV_EXCL_STOP
-  if (Q1d > 10)
-    // LCOV_EXCL_START
-    return CeedError(ceed, CEED_ERROR_BACKEND,
-                     "Magma backend does not support tensor bases with more than 10 quadrature points in each dimension");
-  // LCOV_EXCL_STOP
-
   Ceed_Magma *data;
   ierr = CeedGetData(ceed, &data); CeedChkBackend(ierr);
 
+  // Compile kernels
+  char *magma_common_path;
+  char *interp_path, *grad_path, *weight_path;
+  char *basis_kernel_source;
+  ierr = CeedGetJitAbsolutePath(ceed,
+                                "ceed/jit-source/magma/magma_common_device.h",
+                                &magma_common_path); CeedChkBackend(ierr);
+  CeedDebug256(ceed, 2, "----- Loading Basis Kernel Source -----\n");
+  ierr = CeedLoadSourceToBuffer(ceed, magma_common_path,
+                                &basis_kernel_source);
+  CeedChkBackend(ierr);
+  char *interp_name_base = "ceed/jit-source/magma/interp";
+  CeedInt interp_name_len = strlen(interp_name_base) + 6;
+  char interp_name[interp_name_len];
+  snprintf(interp_name, interp_name_len, "%s-%dd.h", interp_name_base, dim);
+  ierr = CeedGetJitAbsolutePath(ceed, interp_name, &interp_path);
+  CeedChkBackend(ierr);
+  ierr = CeedLoadSourceToInitializedBuffer(ceed, interp_path,
+         &basis_kernel_source);
+  CeedChkBackend(ierr);
+  char *grad_name_base = "ceed/jit-source/magma/grad";
+  CeedInt grad_name_len = strlen(grad_name_base) + 6;
+  char grad_name[grad_name_len];
+  snprintf(grad_name, grad_name_len, "%s-%dd.h", grad_name_base, dim);
+  ierr = CeedGetJitAbsolutePath(ceed, grad_name, &grad_path);
+  CeedChkBackend(ierr);
+  ierr = CeedLoadSourceToInitializedBuffer(ceed, grad_path,
+         &basis_kernel_source);
+  CeedChkBackend(ierr);
+  char *weight_name_base = "ceed/jit-source/magma/weight";
+  CeedInt weight_name_len = strlen(weight_name_base) + 6;
+  char weight_name[weight_name_len];
+  snprintf(weight_name, weight_name_len, "%s-%dd.h", weight_name_base, dim);
+  ierr = CeedGetJitAbsolutePath(ceed, weight_name, &weight_path);
+  CeedChkBackend(ierr);
+  ierr = CeedLoadSourceToInitializedBuffer(ceed, weight_path,
+         &basis_kernel_source);
+  CeedChkBackend(ierr);
+  CeedDebug256(ceed, 2,
+               "----- Loading Basis Kernel Source Complete! -----\n");
+  // The RTC compilation code expects a Ceed with the common Ceed_Cuda or Ceed_Hip
+  // data
+  Ceed delegate;
+  ierr = CeedGetDelegate(ceed, &delegate); CeedChkBackend(ierr);
+  ierr = MAGMA_RTC_COMPILE(delegate, basis_kernel_source, &impl->module, 5,
```

Comment:
The compilation functions get `Ceed_Cuda` or `Ceed_Hip` backend data explicitly, for device properties (like ID), so we can't just pass a `Ceed` whose backend data is `Ceed_Magma`.  This seems to work, but am I breaking any rules here?  Do we need a different solution?

Reply:
This is consistent with what we do with other backends now, though eventually we really should decide how we want backends to reach into each other's data and headers.